pool.go

```diff
@@ -0,0 +1,124 @@
+package go_pool
+
+import (
+	"errors"
+	"runtime"
+	"sync"
+	"time"
+)
+
+var (
+	ErrWorkerClosed = errors.New("no routine is active")
+	ErrJobTimedOut  = errors.New("job request timed expired")
+
+	once    sync.Once
+	poolVar *Pool
+)
+
+// Pool is a struct which contains the list of routines
+type Pool struct {
+	WorkerFun func() Worker
+
+	ReqChan chan interface{}
+	RetChan chan interface{}
+}
+
+// Worker is an interface representing the routine agent
+type Worker interface {
+	Process(interface{}) interface{}
+}
+
+// Initialize is the public function which creates the pool
+func Initialize(nCpus int, nRoutines int, f func(interface{}) interface{}) *Pool {
+	setCpuToBeUsed(nCpus)
+	return New(nRoutines, func() Worker {
+		return &initWorker{
+			processor: f,
+		}
+	})
+}
+
+func New(n int, wFun func() Worker) *Pool {
+	once.Do(func() {
+		poolVar = &Pool{
+			WorkerFun: wFun,
+			ReqChan:   make(chan interface{}, n),
+			RetChan:   make(chan interface{}, n),
+		}
+		go poolVar.initWorkers(n)
+	})
+	return poolVar
+}
+
+func (p *Pool) initWorkers(n int) {
+	var wg sync.WaitGroup
+	for i := 0; i < n; i++ {
+		wg.Add(1)
+		go p.worker(&wg, p.WorkerFun())
```

Comment:
Why spawn all workers at the same time, why not dynamically spawn them when required and make sure max active go routines are capped at n.

Reply:
Dynamically spawning the go routines as and when required by maintaining the counter. Ref - [Commit 1a3db2c](https://github.com/nitinks-dnz/go-pool/pull/1/commits/1a3db2c0366f77d53520f51be49b7be6704041a2)